src/__tests__/integration/data/kv-storage-range.js

```diff
@@ -0,0 +1,74 @@
+export async function handle(state, action) {
+  const input = action.input;
+  const caller = action.caller;
+
+  if (input.function === 'mint') {
+    await SmartWeave.kv.put('mint.' + input.target, input.qty);
+    await SmartWeave.kv.put(input.target, input.qty);
+
+    return {state};
+  }
+
+  if (input.function === 'transfer') {
+    const target = input.target;
+    const qty = input.qty;
+
+    if (!Number.isInteger(qty)) {
+      throw new ContractError('Invalid value for "qty". Must be an integer');
+    }
+
+    if (!target) {
+      throw new ContractError('No target specified');
+    }
+
+    if (qty <= 0 || caller === target) {
+      throw new ContractError('Invalid token transfer');
+    }
+
+    let callerBalance = await SmartWeave.kv.get(caller);
+    callerBalance = callerBalance ? callerBalance : 0;
+
+    if (callerBalance < qty) {
+      throw new ContractError(`Caller balance not high enough to send ${qty} token(s)!`);
+    }
+
+    // Lower the token balance of the caller
+    callerBalance -= qty;
+    await SmartWeave.kv.put(caller, callerBalance);
+
+    let targetBalance = await SmartWeave.kv.get(target);
+    targetBalance = targetBalance ? targetBalance : 0;
+
+    targetBalance += qty;
+    await SmartWeave.kv.put(target, targetBalance);
+
+    return {state};
+  }
+
+  if (input.function === 'balance') {
+    const target = input.target;
+    const ticker = state.ticker;
+
+    if (typeof target !== 'string') {
+      throw new ContractError('Must specify target to get balance for');
+    }
+
+    const result = await SmartWeave.kv.get(target);
+
+    return {result: {target, ticker, balance: result ? result : 0}};
+  }
+
+  if (input.function === 'minted') {
+
+    for (const entry of await await SmartWeave.kv.entries()) {
+      console.log(`getting value ${entry.value} for key ${entry.key}`)
+    }
+
+    const sumMinted = (await SmartWeave.kv.entries({ gte: 'mint.', lte: 'mint.\xff'}))
```

Comment:
on 2nd. thought - 'optimistic' writes and removing values on rollbacks is probably a bad solution, e.g.:

1. A.x makes a 'write' on B
2. B.x (i.e. the 'write' from point 1.) makes a 'put' in KV Storage
3. A.x makes another write on B
3. B.x. (i.e. the write from point 3) makes another 'put' in KV Storage (with the same 'key' as in point 2).
4. Now B.x fails
5. A.x makes a viewState on B, which checks the value of the key set in points 2. and 3.

If we would decide to remove the values, the rollback from point 4. would cause the values for the key and for the 'x' sortKey to be removed completelty - i.e. the viewState would remove null - where it _should_ return value that was set in point 2.